job_runner.py

```diff
@@ -0,0 +1,97 @@
+        print_yellow("#{:^88}#".format(f'Failed - {self.result["failed"]} ({failed_tests})'))
+        print_yellow("#{:^88}#".format(f'Skipped - {self.result["skipped"]} ({skipped_tests})'))
+        print_yellow("#{:^88}#".format(""))
+        print_yellow("{:#^90}".format(""))
+
+    def validate_job(self, yaml_data, yaml_file):
+        schema_path = os.path.join(self.path, "tests", "template", "test_template_schema.json")
+        with open(schema_path, "r") as f:
+            json_schema = yaml.safe_load(f)
+
+        try:
+            validate(yaml_data, json_schema)
+            print(f'Test YAML file valid for {yaml_file}')
+        except ValidationError as err:
+            raise ComplianceException(f"YAML schema validation error - {yaml_file}. Error details - {err.message}")
+
+    def validate_yaml(self, file, yaml_file):
+        try:
+            yaml_data = yaml.safe_load(file)
+            return yaml_data
+        except yaml.YAMLError as err:
+            raise ComplianceException(f"Invalid YAML File - {yaml_file}. Error details - {err}")
+
+    def run_jobs(self):
+
+        yaml_path = os.path.join(self.path, "tests")
+        for yaml_file in os.listdir(yaml_path):
+            if yaml_file.endswith(".yml"):
+                self.test_count += 1
+                with open(os.path.join(yaml_path, yaml_file), "r") as f:
+                    print("\n{:#^100}".format(f"     Initiating Test-{self.test_count} for {yaml_file}     "))
+                    try:
+                        yaml_data = self.validate_yaml(f, yaml_file)
+                        self.validate_job(yaml_data, yaml_file)
+
+                        tag_matched = False
+                        # print(f' {self.tags}   {yaml_data["tags"]}')
+                        for tag in self.tags:
+                            if tag in yaml_data["tags"]:
+                                tag_matched = True
+                                test_runner = TestRunner(yaml_data["server"], yaml_data["version"][0])
+                                job_count = 0
+                                for job in yaml_data["jobs"]:
+                                    job_count += 1
+                                    print(f'Running tests for sub-job-{job_count} -> {job["name"]}')
+                                    test_runner.run_tests(job)
+                                self.result["passed"] += 1
+                                self.test_status["passed"].append(str(self.test_count))
+                                print_green(f'Compliance Test-{self.test_count} for {yaml_file} successful.')
+                        if not tag_matched:
+                            self.result["skipped"] += 1
+                            self.test_status["skipped"].append(str(self.test_count))
+                            print_blue(f"No Tag matched. Skipping Test-{self.test_count} for {yaml_file}")
+
+                    except ComplianceException as err:
```

Comment:
Not so good to have a lot of statements behind `try` - and have a blanket `catch` for all of them. Users will have a hard time telling what went wrong. Seems to me that only a few of the statements in the `try` block would be able to raise a `ComplianceException`.

Reply:
I am catching all the exceptions under one big exception because we have multiple test files, so we run through all files via loop. If a file gives exception, I should still continue on remaining files. So, the exception will just print some error message and update test status instead of exiting the code. The ways to do this are - 
1. Use `continue` while catching, but we cant use this for calling methods, as they are outside loop (eg. .validate_job() or the Test Runner methods). 
2. I use a big exception to catch all of them and print the error in this big exception (i.e. `ComplianceException`)

Please suggest if I can take use any other way

Reply:
Okay, I get it now. But still I have a couple of comments:

1. `ComplianceException` can be misleading. If a server fails a more functional test, e.g., a TES instance is unable to deal with `s3`, it's not a compliance issue, at least not necessarily. How about `TestFailureException`?
2. I don't think that lumping together job validation issues (i.e., problems with the YAML files) with compliance/functional issues of the tested implementations makes a whole lot of sense. If the test files are buggy, it's not the fault of the service being tested, and it's not their responsibility to fix them. So I think those two types of problems should be separated. You could define `JobValidationException` for job validation issues. You may also want to define `TestRunnerException` for issues of the the test runner exception itself (although you can add that at a later point, once the logic is more stable). 
3. Due to what I wrote in the last point, I think that solution (1) you are suggesting, i.e., wrapping only the `test_runner.run_tests(job)` statement in a `try` block and amending the `except` block with `continue, indeed makes more sense here for test failure issues. Then catch the validation errors in a different block and handle accordingly.

Reply:
I have made 3 Exceptions in the suite. `JobValidationException`, `TestFailureException`, `TestRunnerException`. The last one is not used currently, we will add it later.

Regarding that `continue` approach, except the `.validate_yaml()`, all the errors have to be caught under `JobValidationException` & `TestFailureException`. We will be writing the whole `continue` code for just 1 function, which can also be caught under `JobValidationException`.

Should we skip the `continue` method then? So basically, the changes would be, instead of using `ComplianceException` where we do not get a clear message about the exception, we split into 3 well-defined exceptions (`JobValidationException`, `TestFailureException` & `TestRunnerException`), but they will be caught in the same way as before.

Reply:
Let's discuss this on the call now.